File: code_snippets/ts/src/use_cases/async_tasks/simple_async_task/async_task_service.ts
// --------------- define async task logic as a service handler ---------------

import * as restate from "@restatedev/restate-sdk";
import { Context } from "@restatedev/restate-sdk";

// <start_here>
const asyncTaskService = restate.service({
  name: "taskWorker",
  handlers: {
    // <mark_1>
    runTask: async (ctx: Context, params: TaskOpts) => {
      return someHeavyWork(params);
    },
    // </mark_1>
  },
});

export type AsyncTaskService = typeof asyncTaskService;
// <end_here>

restate.endpoint().bind(asyncTaskService).listen(9080);


// ----------------------- Stubs to please the compiler -----------------------

export type TaskOpts = {};

function someHeavyWork(work: TaskOpts) {
  return "Work!";
}

File: code_snippets/python/src/use_cases/async_tasks/simple_async_task/async_task_service.py
from typing import TypedDict

import restate
from restate import Service, Context


class TaskOpts(TypedDict):
    id: str


def some_heavy_work(params: TaskOpts):
    return "Work!"


# <start_here>
async_task_service = Service("taskWorker")


# <mark_1>
@async_task_service.handler("runTask")
async def run_task(ctx: Context, params: TaskOpts):
    return some_heavy_work(params)
# </mark_1>

app = restate.app([async_task_service])
# <end_here>